crates/bevy_pbr/src/render/fog.rs

```diff
@@ -0,0 +1,147 @@
+use bevy_app::{App, Plugin};
+use bevy_asset::{load_internal_asset, HandleUntyped};
+use bevy_ecs::{prelude::*, schedule::SystemLabel};
+use bevy_math::{Vec3, Vec4};
+use bevy_reflect::TypeUuid;
+use bevy_render::{
+    extract_component::ExtractComponentPlugin,
+    render_resource::{DynamicUniformBuffer, Shader, ShaderType},
+    renderer::{RenderDevice, RenderQueue},
+    view::ExtractedView,
+    RenderApp, RenderStage,
+};
+
+use crate::{FogFalloff, FogSettings};
+
+/// The GPU-side representation of the fog configuration that's sent as a uniform to the shader
+#[derive(Copy, Clone, ShaderType, Default, Debug)]
+pub struct GpuFog {
+    /// Unsigned int representation of the active fog falloff mode
+    mode: u32,
+    /// Fog color
+    base_color: Vec4,
+    /// The color used for the fog where the view direction aligns with directional lights
+    directional_light_color: Vec4,
+    /// The exponent applied to the directional light alignment calculation
+    directional_light_exponent: f32,
+    /// Allocated differently depending on fog mode.
+    /// See `mesh_view_types.wgsl` for a detailed explanation
+    be: Vec3,
+    /// Allocated differently depending on fog mode.
+    /// See `mesh_view_types.wgsl` for a detailed explanation
+    bi: Vec3,
+}
```

Comment:
Reordering for better packing of data in the uniform. It makes the order of the fields a bit less nice to read, but it could bring some minor performance benefit GPU-side perhaps.
```suggestion
/// The GPU-side representation of the fog configuration that's sent as a uniform to the shader
#[derive(Copy, Clone, ShaderType, Default, Debug)]
pub struct GpuFog {
    /// Fog color
    base_color: Vec4,
    /// The color used for the fog where the view direction aligns with directional lights
    directional_light_color: Vec4,
    /// Allocated differently depending on fog mode.
    /// See `mesh_view_types.wgsl` for a detailed explanation
    be: Vec3,
    /// The exponent applied to the directional light alignment calculation
    directional_light_exponent: f32,
    /// Allocated differently depending on fog mode.
    /// See `mesh_view_types.wgsl` for a detailed explanation
    bi: Vec3,
    /// Unsigned int representation of the active fog falloff mode
    mode: u32,
}
```